### PR TITLE
Paper editing: add "accept" and "reject" dialogs

### PR DESCRIPTION
**Pull request by pferreir**
This includes a major refactoring of the way timeline code is distributed across Indico, so any opinions on that would be welcome.

### CODE REVIEW

File: indico/modules/events/client/js/reviewing/components/TimelineContent.jsx [hunk 1]
```diff
@@ -8,28 +8,24 @@
...
       </React.Fragment>
     );
   });
 }
 
 TimelineContent.propTypes = {
-  revisions: PropTypes.array.isRequired,
-  state: PropTypes.object.isRequired,
+  blocks: PropTypes.array.isRequired,
+  itemComponent: PropTypes.func.isRequired,
```
Comment (ThiefMaster) [inline indico/modules/events/client/js/reviewing/components/TimelineContent.jsx]: `PropTypes.elementType.isRequired`

File: indico/modules/events/editing/client/js/actions.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+export function reviewEditable(eventId, contributionId, type, revision, formData) {
+  const url = getReviewURL({
+    confId: eventId,
+    contrib_id: contributionId,
+    type,
+    revision_id: revision.id,
+  });
+
+  return async dispatch => {
```
Comment (mic4ael) [inline indico/modules/events/editing/client/js/actions.js]: `submitFormAction`?

File: indico/modules/events/editing/client/js/actions.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import editableDetailsURL from 'indico-url:event_editing.api_editable';
+import getReviewURL from 'indico-url:event_editing.api_review_editable';
+
+import {indicoAxios, handleAxiosError} from 'indico/utils/axios';
+import {camelizeKeys} from 'indico/utils/case';
+
+export const SET_LOADING = 'SET_LOADING';
+export const SET_DETAILS = 'SET_DETAILS';
+
+export function loadTimeline(eventId, contributionId, type) {
```
Comment (mic4ael) [inline indico/modules/events/editing/client/js/actions.js]: what about `ajaxAction` here?

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            text={Translate.string('Judge')}
+            direction="left"
+            button
+            floating
+          >
+            <Dropdown.Menu>
+              {judgmentOptions.map(({value, text}) => (
+                <Dropdown.Item
+                  key={value}
+                  onClick={() => {
```
Comment (mic4ael) [inline indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]:
```suggestion
                  onClick={() => setJudgmentModalType(value)}
```

File: indico/modules/events/editing/client/js/components/Timeline/judgment/AcceptRejectForm.jsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+export default function AcceptRejectForm({block, action, setLoading}) {
+  const {eventId, contributionId, editableType} = useSelector(selectors.getStaticData);
+  const dispatch = useDispatch();
+
+  return (
+    <FinalForm
+      onSubmit={async formData => {
+        setLoading(true);
+        await dispatch(
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/judgment/AcceptRejectForm.jsx]: if you use submitFormAction in the action you can use the return value of the dispatch to pass error information from the server back to final-form (and thus show errors related to a specific field next to that field as long as the field names match on client+server)

File: indico/modules/events/editing/client/js/store.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import createReduxStore from 'indico/utils/redux';
+import reducer, {initialState} from './reducer';
+
+export default staticData =>
+  createReduxStore('editing-timeline', {timeline: reducer}, {timeline: initialState, staticData});
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/store.js]: do we need the `timeline: initialState` here? it's already provided in the reducer's default argument

File: indico/modules/events/editing/client/js/actions.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  const url = editableDetailsURL({confId: eventId, contrib_id: contributionId, type});
+  return ajaxAction(
+    async () => {
+      const result = await indicoAxios.get(url);
+      return camelizeKeys(result);
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/actions.js]: Does this actually work? `result` is a Response and not the actual data.
For camelizing data, using the 5th (transformData) arg of ajaxAction would be better

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import UserAvatar from 'indico/modules/events/reviewing/components/UserAvatar';
+import {FinalCheckbox, FinalInput, FinalSubmitButton, FinalTextArea} from 'indico/react/forms';
+import {Translate} from 'indico/react/i18n';
+
+import JudgmentBox from './judgment/JudgmentBox';
+import {blockPropTypes} from './util';
+
+import './ReviewForm.module.scss';
+
+export const judgmentOptions = [
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: I don't think that `export` is needed

File: indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+];
+
+export default function ReviewForm({block}) {
+  const currentUser = {
+    fullName: Indico.User.full_name,
+    avatarBgColor: Indico.User.avatar_bg_color,
+  };
+
+  const [commentFormVisible, setCommentFormVisible] = useState(false);
+  const [judgmentModalType, setJudgmentModalType] = useState(null);
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/ReviewForm.jsx]: It's not a Modal anymore

**Comment by mic4ael** [comment]
Still WIP in the last commit message ;)

**Comment by pferreir** [comment]
> Still WIP in the last commit message ;)

Nitpicking! 😉
Sorry, fixed.

File: indico/modules/events/editing/client/js/components/Timeline/judgment/AcceptRejectForm.jsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        setLoading(true);
+        const ret = await dispatch(
+          reviewEditable(eventId, contributionId, editableType, block, {...formData, action})
+        );
+        if (ret.error) {
+          setLoading(false);
+          return ret.error;
+        }
+      }}
+      subscription={{submitting: true}}
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/judgment/AcceptRejectForm.jsx]: you aren't using `submitting` anymore, so specify an empty object instead

File: indico/modules/events/editing/client/js/components/Timeline/judgment/JudgmentBox.jsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import AcceptRejectForm from './AcceptRejectForm';
+
+import './JudgmentBox.module.scss';
+
+export default function JudgmentBox({block, onClose, judgmentType: _judgmentType, options}) {
+  const [judgmentType, setJudgmentType] = useState(_judgmentType);
+  const [loading, setLoading] = useState(false);
+  const optionsMap = useMemo(() => _.keyBy(options, 'value'), [options]);
+
+  const option = optionsMap[judgmentType];
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/components/Timeline/judgment/JudgmentBox.jsx]: Can't this be simplified to

    const option = options.find(x => x.value === judgmentType);

File: indico/modules/events/editing/client/js/reducer.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import {SET_LOADING, SET_DETAILS} from './actions';
+import {processRevisions} from './selectors';
+
+export const initialState = {
+  details: null,
+  timelineBlocks: null,
+  isLoading: false,
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/reducer.js]: just `loading`?

File: indico/modules/events/editing/client/js/selectors.js [hunk 13]
```diff
@@ -118,3 +119,22 @@ export function processRevisions(revisions) {
...
+);
+export const getLastRevision = createSelector(
+  getDetails,
+  details => details && details.revisions[details.revisions.length - 1]
+);
+export const getLastState = createSelector(
+  getLastRevision,
+  lastRevision =>
+    lastRevision &&
+    (lastRevision.finalState.name === 'none' ? lastRevision.initialState : lastRevision.finalState)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/client/js/selectors.js]: Let's use the enum here (ie `FinalRevisionState.none`)?

Comment (mic4ael) [reply indico/modules/events/editing/client/js/selectors.js]: There is already an enum for that ;)

**Pull request merged**